README.md
# Hoodie Integration Tests
[![Build Status](https://travis-ci.org/hoodiehq/hoodie-integration-test.svg)](https://travis-ci.org/hoodiehq/hoodie-integration-test)
[![Dependency Status](https://david-dm.org/hoodiehq/hoodie-integration-test.svg)](https://david-dm.org/hoodiehq/hoodie-integration-test)
[![devDependency Status](https://david-dm.org/hoodiehq/hoodie-integration-test/dev-status.svg)](https://david-dm.org/hoodiehq/hoodie-integration-test#info=devDependencies)

*NOTE: When running this locally make sure to download and start the [Selenium Server](http://www.seleniumhq.org/download/) first*
*NOTE: currently this package has to be consumed via [grunt-subgrunt](https://github.com/tusbar/grunt-subgrunt).*

This repo includes a test scenario that mirrors the Hoodie end-user-experience from install & setup to first run. It exists to ensure that Hoodie is usable at all times by all users.

Tests can be run on [Travis](https://travis-ci.org) and/or locally.

```js
require('hoodie-integration-test')();
```

That's all (:

Should you be prompted to enter credentials like so `Please enter your CouchDB _admin credentials:` kill all running CouchDB processes.
```bash
ps ax | grep couch | awk '{print $1}' | xargs kill
``` 

## Local setup

```bash
grunt dev # creates a hoodie app for local debugging
```


CHANGELOG.md
<a name"1.0.9"></a>
### 1.0.9 (2015-07-08)


<a name"1.0.8"></a>
### 1.0.8 (2015-07-03)


#### Bug Fixes

* stringify error before passing to test callback (27a6aaa2)
* **env:** manually add the test app to the paths that require looks up (cc178d5f)


<a name="1.0.7"></a>
### 1.0.7 (2015-03-01)


#### Bug Fixes

* dependency map for deep linking for hoodie-plugin-users ((7a55454c))


<a name="1.0.4"></a>
### 1.0.4 (2015-02-21)


#### Bug Fixes

* double maxBuffer to avoid fails on travis ((3b788a62))


<a name="1.0.0"></a>
## 1.0.0 (2014-10-22)


#### Bug Fixes

* **plugin:** install in correct location ((9cca5055))


<a name="0.3.2"></a>
### 0.3.2 (2014-10-14)


<a name="0.3.1"></a>
### 0.3.1 (2014-10-13)


#### Bug Fixes

* **utils:** use latest selenium server ((ca5420b2))


<a name="0.3.0"></a>
## 0.3.0 (2014-10-08)


<a name="0.3.0-beta.1"></a>
### 0.3.0-beta.1 (2014-09-23)


#### Bug Fixes

* bearer token support ((28eb1bba))
* run both www & admin tests ((83c82bd4))
* **api:** task tests ((c5ef928b))


<a name="0.2.1"></a>
### 0.2.1 (2014-09-17)


#### Bug Fixes

* **integration-test:** remove testapp afterwards so it is not stored in travis cache ((d6a0ad6a))


<a name="0.2.0"></a>
## 0.2.0 (2014-09-06)


<a name="0.2.0-rc.2"></a>
### 0.2.0-rc.2 (2014-08-21)


#### Bug Fixes

* **integration-test:** use child_process.exec for cwd to have an effect ((d52f989c))


<a name="0.2.0-rc.1"></a>
### 0.2.0-rc.1 (2014-08-21)


#### Bug Fixes

* make test more robust by replacing wait with findTimeout ((42ea8bec))
* inject Function.prototype.bind polyfill into the app ((267b0ccc))
* **test:** inject bind polyfill to `grunt test` task ((19b47448))


<a name="0.1.4"></a>
### 0.1.4 (2014-08-07)


<a name="0.1.3"></a>
### 0.1.3 (2014-08-06)


#### Bug Fixes

* **test:** app needs to be created before npm can list deps ((70717de1))


<a name="0.1.2"></a>
### 0.1.2 (2014-08-05)


#### Bug Fixes

* **test:** do not test if module is not in dependency list ((193d8a9e))


<a name="0.1.1"></a>
### 0.1.1 (2014-08-05)


<a name="0.1.0"></a>
## 0.1.0 (2014-08-05)


<a name="0.0.4"></a>
### 0.0.4 (2014-07-21)


#### Bug Fixes

* **createApp:** remove cache clean as it is deprecated ((3b5ec295))


<a name="0.0.3"></a>
### 0.0.3 (2014-07-17)


#### Bug Fixes

* require function call to run tests ((fadca7fc))


<a name="0.0.2"></a>
### 0.0.2 (2014-07-17)


#### Features

* add main script to make this require-able ((adcf3cde))


<a name="0.0.1"></a>
### 0.0.1 (2014-07-17)




test.sh
#!/bin/sh -e

MAIN_STATUS=`curl -s http://127.0.0.1:6001 -w "\n%{http_code}\n" | tail -1`
MAIN_OK=$?

API_STATUS=`curl -s http://127.0.0.1:6001/_api -w "\n%{http_code}\n" | tail -1`
API_OK=$?

ADMIN_STATUS=`curl -s http://127.0.0.1:6002 -w "\n%{http_code}\n" | tail -1`
ADMIN_OK=$?

COUCH_STATUS=`curl -s http://127.0.0.1:6003 -w "\n%{http_code}\n" | tail -1`
COUCH_OK=$?

if [ $MAIN_OK -ne 0 ]; then
  echo "MAIN EXIT STATUS NOT 0"
  exit 1
fi

if [ "$MAIN_STATUS" != "200" ]; then 
  echo "MAIN HTTP STATUS NOT 200"
  exit 2
fi


if [ $API_OK -ne 0 ]; then
  echo "API EXIT STATUS NOT 0"
  exit 3
fi

if [ "$API_STATUS" != "200" ]; then 
  echo "API HTTP STATUS NOT 200"
  exit 4
fi


if [ $ADMIN_OK -ne 0 ]; then
  echo "ADMIN EXIT STATUS NOT 0"
  exit 5
fi

if [ "$ADMIN_STATUS" != "200" ]; then 
  echo "ADMIN HTTP STATUS NOT 200"
  exit 6
fi


if [ $COUCH_OK -ne 0 ]; then
  echo "COUCH EXIT STATUS NOT 0"
  exit 7
fi

if [ "$COUCH_STATUS" != "200" ]; then 
  echo "COUCH HTTP STATUS NOT 200"
  exit 8
fi

echo "All Good! :)"


tests/test.js
define([
  'intern!object',
  'intern/chai!',
  'intern/dojo/node!chai-as-promised',
  'intern/chai!expect',
  'intern/dojo/node!../hosts.json',
  'intern/dojo/node!./www',
  'intern/dojo/node!./admin'
], function (registerSuite, chai, chaiAsPromised, expect, hosts, www, admin) {
  chai.use(chaiAsPromised);
  chai.should();
  registerSuite(www(expect, hosts));
  registerSuite(admin(expect, hosts));
});


tests/pre-test.js
/*global casper:false */

casper.options.viewportSize = {width: 1024, height: 768};
casper.test.comment('pre');
casper.on('remote.message', function(message) {
  console.log('CONSOLE', message);
});
casper.on('remote.alert', function(message) {
  console.log('ALERT', message);
});
casper.on('page.error', function(message, trace) {
  console.log(message);
  console.log(trace);
});
casper.test.done();


util/dep-path.js
var exec = require('child_process').exec;
var path = require('path');

var NM = 'node_modules';

var exports = module.exports = function(cwd, module, callback) {
  exports.getNpmList(cwd, function(list) {
    if (!exports.handleDependencies(module, './', list.dependencies, callback)) {
      callback(null);
    }
  });
};

// workaround for https://github.com/hoodiehq/hoodie-integration-test/issues/19
// map module names to paths where we have to run `npm link <moduleName>` in
exports.dependenciesMap = {
  'hoodie-server': '.',
  'hoodie': 'node_modules/hoodie-server',
  'hoodie-admin-dashboard': 'node_modules/hoodie-server',
  'hoodie-admin-dashboard-uikit': 'node_modules/hoodie-server',
  'hoodie-plugins-manager': 'node_modules/hoodie-server',
  'hoodie-plugins-api': 'node_modules/hoodie-server/node_modules/hoodie-plugins-manager',
  'hoodie-plugin-appconfig': '.',
  'hoodie-plugin-email': '.',
  'hoodie-plugin-users': '.'
}

exports.getNpmList = function(cwd, callback) {
  exec('npm ls --json --depth=3', {
    cwd: path.resolve(cwd) || process.cwd()
  }, function(err, stdout) {
    try {
      callback(JSON.parse(stdout));
    } catch (_e) {
      if (err) {
        throw err;
      }
      throw _e;
    }
  });
};

exports.handleDependencies = function(module, depPath, deps, callback) {
  // This will only find the first occurance of the module
  // which is entirely cool for hoodie's usecase
  /*jshint forin: false */
  for (var name in deps) {
    var dep = deps[name];
    if (name === module) {
      callback(depPath);
      return true;
    }
    if (dep.dependencies) {
      if (exports.handleDependencies(
        module,
        path.join(depPath, NM, name),
        dep.dependencies,
        callback
      )) {
        return true;
      }
    }
  }
  return false;
};



Gruntfile.js
var path = require('path');

var cp = require('child_process');
var shell = require('shelljs');

var depPath = require('./util/dep-path');

module.exports = function(grunt) {
  'use strict';

  // load npm tasks
  require('load-grunt-tasks')(grunt);
  grunt.loadNpmTasks('intern');

  var env = process.env;
  env.HOODIE_SETUP_PASSWORD = '12345';
  var ports = require('./ports');
  var appname = 'myapp';

  // Project configuration.
  grunt.initConfig({
    watch: {
      dev: {
        files: ['tests/**/*'],
        tasks: [
          'rm-data',
          'hoodie:start',
          'intern:tests',
          'hoodie:stop',
        ]
      },
      gruntfile: {
        files: ['Gruntfile.js']
      }
    },

    hoodie: {
      start: {
        options: {
          childProcessOptions: {
            cwd: process.cwd() + '/myapp',
            env: env
          },
          'custom-ports': [ports.www, ports.admin, ports.couch].join(','),
          callback: function(config) {
            var output = {};
            Object.keys(config.stack).forEach(function(type) {
              if (config.stack[type].port !== ports[type]) {
                grunt.fail.fatal('Hoodie not running on custom ports');
              }
              grunt.log.ok('Hoodie running on correct ' + type + ' port: ' + ports[type]);
              output[type] = 'http://' +
                config.stack[type].host +
                ':' +
                config.stack[type].port +
                '/';
            });
            var fs = require('fs');
            fs.writeFileSync('./hosts.json', JSON.stringify(output, null, 2));
            grunt.log.writeflags(ports, 'Hoodie running on correct ports');
          }
        }
      },
      stop: {}
    },

    shell: {
      options: {
        execOptions: {
          // https://www.npmjs.com/package/grunt-shell#execoptions
          maxBuffer: 400*1024
        }
      },
      createApp: {
        command: './node_modules/.bin/hoodie new ' + appname
      },
      installTestPlugin: {
        command: 'cd ' + appname + ' && ./../node_modules/.bin/hoodie install test && cd -'
      }
    },

    intern: {
      options: {
        config: 'intern.js',
        runType: 'runner'
      },
      tests: {}
    }
  });

  grunt.registerTask('install-selenium', function() {
    shell.exec('sh util/install-selenium.sh');
  });

  grunt.registerTask('rm-app', function() {
    shell.rm('-rf', 'myapp');
  });

  grunt.registerTask('rm-data', function() {
    shell.rm('-rf', 'myapp/data');
  });

  // Default task(s).
  grunt.registerTask('dev', [
    'rm-app',
    'shell:createApp',
    'shell:installTestPlugin',
    'watch'
  ]);

  grunt.registerTask('test', function() {
    var module = this.args.join('');
    var tasksPre = ['shell:createApp', 'shell:installTestPlugin'];
    var tasksPost = ['hoodie:start', 'intern:tests', 'hoodie:stop', 'rm-app'];

    if (process.env.CI) {
      tasksPre.push('install-selenium');
    }

    if (!module) {
      return grunt.task.run(tasksPre.concat(tasksPost));
    }

    grunt.registerTask('deep-link', function() {
      var done = this.async();
      var modulePath = depPath.dependenciesMap[module];

      if (!modulePath) {
        grunt.log.warn('Dependencies do not contain the module ' + module);
        return done();
      }

      cp.exec('npm link ' + module, {
        cwd: path.resolve(path.join(appname, modulePath)),
        maxBuffer: 400*1024
      }, function(err, stdout, stderr) {
        if (err) {
          console.log('execution error: ', err);
        }
        console.log(stdout);
        console.log(stderr);
        grunt.task.run(tasksPost);
        done();
      });

      // see https://github.com/hoodiehq/hoodie-integration-test/issues/19
      // depPath(appname, module, function(depPath) {
      //   if (!depPath) {
      //     grunt.log.warn('Dependencies do not contain the module ' + module);
      //     return done();
      //   }

      //   cp.exec('npm link ' + module, {
      //     cwd: path.resolve(path.join(appname, depPath)),
      //     maxBuffer: 400*1024
      //   }, function(err, stdout, stderr) {
      //     if (err) {
      //       console.log('execution error: ', err);
      //     }
      //     console.log(stdout);
      //     console.log(stderr);
      //     grunt.task.run(tasksPost);
      //     done();
      //   });
      // });
    });

    grunt.task.run(tasksPre.concat(['deep-link']));
  });

  grunt.registerTask('default', ['rm-app', 'test']);
};

